src/System.Memory/src/System/ReadOnlySpan.cs

```diff
@@ -82,12 +81,11 @@ public ReadOnlySpan(T[] array, int start)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public ReadOnlySpan(T[] array, int start, int length)
         {
-            if (array == null)
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
-            if (default(T) == null && array.GetType() != typeof(T[]))
+            if (array == null || (start | length) < 0 || array.Length - start < length)
+                ThrowHelper.ThrowArgumentNullOrOutOfRangeException(array);
+            // Check will be Jitted out for ValueTypes
+            if (SpanHelpers.IsReference<T>() && array.GetType() != typeof(T[]))
                 ThrowHelper.ThrowArrayTypeMismatchException_ArrayTypeMustBeExactMatch(typeof(T));
-            if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
```

Comment:
Though I did notice something strange about the rest of the coreclr method https://github.com/dotnet/coreclr/issues/9633